File: src/main/java/net/sharplab/epubtranslator/core/driver/epub/EPubWriter.java
package net.sharplab.epubtranslator.core.driver.epub;

import net.sharplab.epubtranslator.core.model.EPubContentFile;
import net.sharplab.epubtranslator.core.model.EPubFile;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class EPubWriter {
    public void write(EPubFile ePubFile, File dstFile) {
        try(FileOutputStream fileOutputStream = new FileOutputStream(dstFile);
            ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)){
            for(EPubContentFile contentFile : ePubFile.getContentFiles()){
                ZipEntry zipEntry = new ZipEntry(contentFile.getName());
                zipOutputStream.putNextEntry(zipEntry);
                zipOutputStream.write(contentFile.getData());
            }
        }
        catch (IOException ex) {
            throw new UncheckedIOException(ex);
        }
    }
}

File: src/main/java/net/sharplab/epubtranslator/core/driver/epub/EPubWriter.kt
package net.sharplab.epubtranslator.core.driver.epub

import net.sharplab.epubtranslator.core.model.EPubFile
import java.io.File
import java.io.FileOutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import java.util.zip.CRC32

class EPubWriter {
    fun write(ePubFile: EPubFile, dstFile: File) {
        FileOutputStream(dstFile).use { fileOutputStream ->
            ZipOutputStream(fileOutputStream).use { zipOutputStream ->
                val mimetypeEntry = ZipEntry("mimetype")

                mimetypeEntry.method = ZipEntry.STORED
                val mimetypeBytes = "application/epub+zip".toByteArray()
                mimetypeEntry.size = mimetypeBytes.size.toLong()

                val crc = CRC32()
                crc.update(mimetypeBytes, 0, mimetypeBytes.size)
                mimetypeEntry.crc = crc.value

                zipOutputStream.putNextEntry(mimetypeEntry)
                zipOutputStream.write("application/epub+zip".toByteArray())

                ePubFile.contentFiles.forEach {
                    if (it.name != "mimetype") { // mimetypeはすでに追加されているのでスキップ
                        val zipEntry = ZipEntry(it.name)
                        zipOutputStream.putNextEntry(zipEntry)
                        zipOutputStream.write(it.data)
                    }
                }
            }
        }
    }
}